c_api/tests/mpm88_test.cpp

```diff
@@ -0,0 +1,237 @@
+
+    v_ = NdarrayAndMem::Make(runtime_, TiDataType::TI_DATA_TYPE_F32,
+                             {kNrParticles}, vec2_shape);
+
+    pos_ = NdarrayAndMem::Make(runtime_, TiDataType::TI_DATA_TYPE_F32,
+                               {kNrParticles}, vec3_shape, false, false);
+
+    C_ = NdarrayAndMem::Make(runtime_, TiDataType::TI_DATA_TYPE_F32,
+                             {kNrParticles}, mat2_shape);
+
+    J_ = NdarrayAndMem::Make(runtime_, TiDataType::TI_DATA_TYPE_F32,
+                             {kNrParticles}, {});
+
+    grid_v_ = NdarrayAndMem::Make(runtime_, TiDataType::TI_DATA_TYPE_F32,
+                                  {kNGrid, kNGrid}, vec2_shape);
+    grid_m_ = NdarrayAndMem::Make(runtime_, TiDataType::TI_DATA_TYPE_F32,
+                                  {kNGrid, kNGrid}, {});
+
+    k_init_particles_ = ti_get_aot_module_kernel(module_, "init_particles");
+    k_substep_g2p_ = ti_get_aot_module_kernel(module_, "substep_g2p");
+    k_substep_reset_grid_ =
+        ti_get_aot_module_kernel(module_, "substep_reset_grid");
+    k_substep_p2g_ = ti_get_aot_module_kernel(module_, "substep_p2g");
+    k_substep_update_grid_v_ =
+        ti_get_aot_module_kernel(module_, "substep_update_grid_v");
+
+    k_init_particles_args_[0] = x_->argument();
+    k_init_particles_args_[1] = v_->argument();
+    k_init_particles_args_[2] = J_->argument();
+
+    k_substep_reset_grid_args_[0] = grid_v_->argument();
+    k_substep_reset_grid_args_[1] = grid_m_->argument();
+
+    k_substep_p2g_args_[0] = x_->argument();
+    k_substep_p2g_args_[1] = v_->argument();
+    k_substep_p2g_args_[2] = C_->argument();
+    k_substep_p2g_args_[3] = J_->argument();
+    k_substep_p2g_args_[4] = grid_v_->argument();
+    k_substep_p2g_args_[5] = grid_m_->argument();
+
+    k_substep_update_grid_v_args_[0] = grid_v_->argument();
+    k_substep_update_grid_v_args_[1] = grid_m_->argument();
+
+    k_substep_g2p_args_[0] = x_->argument();
+    k_substep_g2p_args_[1] = v_->argument();
+    k_substep_g2p_args_[2] = C_->argument();
+    k_substep_g2p_args_[3] = J_->argument();
+    k_substep_g2p_args_[4] = grid_v_->argument();
+    k_substep_g2p_args_[5] = pos_->argument();
+
+    ti_launch_kernel(runtime_, k_init_particles_, 3,
+                     &k_init_particles_args_[0]);
+
+    ti_wait(runtime_);
+  }
+
+  ~MPM88DemoImpl() {
+    ti_destroy_aot_module(module_);
+    ti_destroy_runtime(runtime_);
+  }
+
+  void Step() {
+    for (size_t i = 0; i < N_ITER; i++) {
+      ti_launch_kernel(runtime_, k_substep_reset_grid_, 2,
+                       &k_substep_reset_grid_args_[0]);
+      ti_launch_kernel(runtime_, k_substep_p2g_, 6, &k_substep_p2g_args_[0]);
+      ti_launch_kernel(runtime_, k_substep_update_grid_v_, 2,
+                       &k_substep_update_grid_v_args_[0]);
+      ti_launch_kernel(runtime_, k_substep_g2p_, 6, &k_substep_g2p_args_[0]);
+    }
+    ti_wait(runtime_);
+  }
+
+ private:
+  class NdarrayAndMem {
```

Comment:
for followups: shall we move this into c-api since it's pretty commonly useful? ;)